guillotina/db/strategies/resolve.py

```diff
@@ -1,56 +0,0 @@
-from guillotina import configure
-from guillotina import glogging
-from guillotina.db.interfaces import IDBTransactionStrategy
-from guillotina.db.interfaces import ITransaction
-from guillotina.db.strategies.simple import SimpleStrategy
-
-
-logger = glogging.getLogger("guillotina")
-
-
-@configure.adapter(for_=ITransaction, provides=IDBTransactionStrategy, name="resolve")
-class ResolveStrategy(SimpleStrategy):
-    """
-    If simultaneous transactions are not editing the same objects, let it go
-    """
-
-    async def tpc_vote(self):
-        if not self.writable_transaction:
-            return True
-        # potential conflict error, get changes
-        # Check if there is any commit bigger than the one we already have
-        conflicts = await self._storage.get_conflicts(self._transaction)
-        tids = []
-        for conflict in conflicts:
-            # both writing to same object...
-            tids.append(conflict["tid"])
-            if conflict["zoid"] in self._transaction.modified:
-                modified_keys = [k for k in self._transaction.modified.keys()]
-                logger.warn(
-                    f"Could not resolve conflicts in TID: {self._transaction._tid}\n"
-                    f'Conflicted TID: {conflict["tid"]}\n'
-                    f"IDs: {modified_keys}"
-                )
-                return False
-        if len(conflicts) > 0:
-            logger.info(
-                "Resolved conflict between transaction ids: {}, {}".format(self._transaction._tid, tids)
-            )
-
-        return True
-
-
-@configure.adapter(for_=ITransaction, provides=IDBTransactionStrategy, name="resolve_readcommitted")
-class ResolveReadCommittedStrategy(ResolveStrategy):
-    """
-    Resolve strategy that delays starting transaction to commit phase
-    which means we have shorter db transactions but reads might not be inconsistent.
-    """
-
-    async def tpc_begin(self):
-        pass
-
-    async def tpc_commit(self):
-        await self.retrieve_tid()
-        if self._transaction._db_txn is None:
-            await self._storage.start_transaction(self._transaction)
```

Comment:
@bloodbare I was reviewing this part and I don't understand why the strategy `resolve_readcommitted` (the default one) needs the `tpc_vote` logic from `resolve` strategy.

If I'm not wrong, these are the cases we are trying to mitigate:

1. Cache invalidation race condition: an object is modified but the next request that wants to modify the object reads the stale object. During the write to the storage, we want to detect that the old txn id is not the latest in postgres. This is already implemented in the UPDATE sql, using the `WHERE tid = $7::bigint`. If the WHERE clause doesn't match it would not update the object and we'll raise the ConflictError.

2. Two concurrent requests modify the same object "at the same time". At the same time could mean two things:
a) Both requests read the same object, the first starts the pg transaction and writes the changes and finished. When the second transaction starts will try to update the object and I'd would happen the error in point 1. The Old TID doesn't match and the ConflictError is raised.
b) Both requets read the same object, the first request starts the transaction and writes the changes. At the same time, before committing the pg transaction the second request starts a transaction and tries to write the same object. In that case two things could happen: 1. same thing as previous requests. 2. deadlock and both requests fail

My conclusion is that in all these situations postgres/SQL queries already handle the conflicts, and there is no need for the extra logic. The only case I could think this is useful is for Cockroach DB.

./cc @vangheem

Reply:
Since this is something that didn't change in this PR, I move this discussion to an issue to review it in the future